regen/HeaderParser.pm

```diff
@@ -0,0 +1,1556 @@
+package HeaderParser;
+use strict;
+use warnings;
+
+# these are required below in BEGIN statements, we cant have a
+# hard dependency on them as they might not be available when
+# we run as part of autodoc.pl
+#
+# use Data::Dumper;
+# use Storable qw(dclone);
+#
+use Carp       qw(confess);
+use Text::Tabs qw(expand unexpand);
+use Text::Wrap qw(wrap);
+
+# The style of this file is determined by:
+#
+# perltidy -w -ple -bbb -bbc -bbs -nolq -l=80 -noll -nola -nwls='=' \
+#   -isbc -nolc -otr -kis -ci=4 -se -sot -sct -nsbl -pt=2 -fs  \
+#   -fsb='#start-no-tidy' -fse='#end-no-tidy'
+
+my (
+    %unop,              # unary operators and their precedence
+    %binop,             # binary operators and their precedence
+    %is_right_assoc,    # operators which are right associative
+    %precedence,        # precedence of all operators.
+    %associative,       # associative operators
+    %commutative,       # commutative operators
+    %cmpop,             # comparison operators
+    $unop_pat,          # pattern to match unary operators
+    $binop_pat,         # pattern to match binary operators
+    %op_names,          # map of op to description, used in error messages
+    $tokenize_pat       # a pattern which can tokenize an expression
+);
+
+BEGIN {
+    # this is initialization for the operator precedence expression parser
+    # we use for handling preprocessor conditions.
+    %op_names= (
+        '==' => 'equality',
+        '!=' => 'inequality',
+        '<<' => 'bit-shift-left',
+        '>>' => 'bit-shift-right',
+        '+'  => 'addition',
+        '-'  => 'subtraction',
+        '*'  => 'multiplication',
+        '/'  => 'division',
+        '%'  => 'modulo',
+        '||' => 'logical-or',       # Lowest precedence
+        '&&' => 'logical-and',
+        '|'  => 'binary-or',
+        '^'  => 'binary-xor',
+        '&'  => 'binary-and',
+        '<'  => 'less-than',        # split on spaces, all with equal precedence
+        '>'  => 'greater-than',
+        '<=' => 'less-than-or-equal',
+        '>=' => 'greater-than-or-equal',
+    );
+    my @cmpop= (
+        '== !=',                    # listed in lowest to highest precedence
+        '< > <= >=',                # split on spaces, all with equal precedence
+    );
+    my @binop= (
+        '||',                       # Lowest precedence
+        '&&',
+        '|',
+        '^',
+        '&',
+        @cmpop,    # include the numerical comparison operators.
+        '<< >>',
+        '+ -',
+        '* / %',    # highest prcedence operators.
+    );
+
+    my @unop= qw( ! ~ + - );
+    %unop= map  { $_ => 1 } @unop;
+    %cmpop= map { $_ => 1 } map { split /\s+/, $_ } @cmpop;
+    %binop= map { $_ => 1 } map { split /\s+/, $_ } @binop;
```

Comment:
Minor nit, repeated in many places:  I like a space before the '=' and I think we mostly do that.

Reply:
This comes up every now and then. I left hug assignment operators for my own good reasons. I include perltidy rules in the docs of my files for following my style. We don't have an official style I can use to format other folks code, but I provide one so folks can use it to format changes to my code. I also "follow the style of the file" when I am modifying existing files. I think that is a reasonable compromise.  

FWIW, I would welcome us adopting a standard perltidy ruleset for our test files and porting code and the like. But until then I would prefer that files I author and will maintain be in the style I prefer to use and in which I created them. Call it my artistic license. (Although I have good technical reasons for my preferences. I *never* mistakenly use assignment instead of equality operators.) I don't think that is unreasonable, especially as I would be happy to switch to the "approved" perltidy ruleset if we could agree on it. (There are tons of files I *long* to perltidy.)

Reply:
If there's any standard to follow for code format it would be https://perldoc.perl.org/perlstyle

Reply:
Until we agree on a set of official coding standards such that I can run perltidy with those rules on any file I modify (to fix my mistakes) then I will continue to follow the tradition of the project, which is to respect that others may have differing style choices and to follow the style of the file, granting the author of the file the choice of its style.

Again, I would vote in favor of any sane set of perltidy rules to mechanize this process. FWIW, I feel some of the guidance on the perltstyle doc is controversial and at least shouldn't be part of a style guide /unless/ it can be machine assisted.

I am in fact in favor of any machine assisted style management, indeed this patch is intended to introduce machine assisted tidying to embed.fnc itself. :-) So long as I can run a tool to fix things I dont care what we use, until then I'd prefer the things I author and intend to maintain follow my preferences. At least when I work on those I don't have to worry about my style, any more than anybody else does, as I just run my perltidy rules afterwards.